examples/bridge-app/linux/UserInputBackend.cpp

```diff
@@ -0,0 +1,328 @@
+    }
+    else
+    {
+        printf("Unknown device type %s\n", tokens[1].c_str());
+        return;
+    }
+}
+
+void AddCluster(const std::vector<std::string> & tokens)
+{
+    if (!g_pending)
+    {
+        printf("Expected pending device!\n");
+        return;
+    }
+    if (tokens.size() < 2)
+    {
+        printf("Expected id of a device type\n");
+        return;
+    }
+
+    for (const auto & cluster : clusters::kKnownClusters)
+    {
+        if (tokens[1] == cluster.name)
+        {
+            std::unique_ptr<CommonCluster> obj(cluster.ctor(::operator new(cluster.size)));
+            DynamicAttributeList dynamic_attrs;
+            auto attrs = obj->GetAllAttributes();
+            for (auto & attr : attrs)
+                dynamic_attrs.Add(attr);
+            g_pending->AddCluster(std::move(obj), dynamic_attrs, nullptr, nullptr);
+        }
+    }
+}
+
+void CancelDevice(const std::vector<std::string> & tokens)
+{
+    g_pending.reset(nullptr);
+}
+
+void FinishDevice(const std::vector<std::string> & tokens)
+{
+    if (!g_pending)
+    {
+        printf("Expected pending device!\n");
+        return;
+    }
+
+    for (auto * c : g_pending->clusters())
+        c->SetCallback(&g_write_cb);
+
+    auto dev = std::make_unique<Device>(g_pending->CreateDevice());
+    int ep   = AddDeviceEndpoint(dev.get());
+    if (ep < 0)
+    {
+        printf("Failed to add device\n");
+    }
+    else
+    {
+        printf("Added device at index %d\n", ep);
+        size_t index = (size_t) ep;
+        if (g_devices.size() <= index)
+        {
+            g_devices.resize(index + 1);
+            g_device_impls.resize(index + 1);
+        }
+        g_devices[index]      = std::move(dev);
+        g_device_impls[index] = std::move(g_pending);
+    }
+}
+
+void RemoveDevice(const std::vector<std::string> & tokens)
+{
+    if (tokens.size() < 2)
+    {
+        printf("Expected index of a device\n");
+        return;
+    }
+    uint32_t index = atoi(tokens[1].c_str());
+    if (index >= g_devices.size())
+    {
+        printf("%d is an invalid index\n", index);
+        return;
+    }
+
+    RemoveDeviceEndpoint(g_devices[index].get());
+
+    g_devices[index]      = nullptr;
+    g_device_impls[index] = nullptr;
+}
+
+void AddType(const std::vector<std::string> & tokens)
+{
+    if (!g_pending)
+    {
+        printf("Expected pending device!\n");
+        return;
+    }
+    if (tokens.size() < 2)
+    {
+        printf("Expected id of a device type\n");
+        return;
+    }
+    EmberAfDeviceType type = { (uint16_t) atoi(tokens[1].c_str()), 1u };
+    if (tokens.size() >= 3)
+    {
+        type.deviceVersion = (uint8_t) atoi(tokens[2].c_str());
+    }
+    printf("Adding device type %d ver %d\n", type.deviceId, type.deviceVersion);
+
+    g_pending->AddDeviceType(type);
+}
+
+bool FindIndexAttr(const std::vector<std::string> & tokens, ClusterImpl ** cluster, chip::AttributeId * attrId)
```

Comment:
Could we separate out the token parsing from the method implementation?

This does both:
  - convert tokens to integers
  - skips the 1st token (so why is it passed in)
  - locate device and attribute(

Please try to add a comment to describe what this does. As I understand it, this is a: `if device impl index denoted by int(arg[1]) exists and contains cluster id denoted by int(arg[2]) then set the cluster impl to the cluster and attribute id to the value of arg[3]`

In that case, why do we do the `*attrId = (chip::AttributeId) atoi(tokens[3].c_str());` in this method? Why is this not a GetClusterByDevice(device_index, cluster_id)?